vochain/ist/ist.go

```diff
@@ -0,0 +1,303 @@
+	// get the IST actions
+	actions := Actions{}
+	actionsBytes, err := c.retrieveFromNoState(dbIndex(height))
+	if err != nil {
+		if errors.Is(err, db.ErrKeyNotFound) {
+			// no IST actions scheduled for this height, we return empty
+			actions.Actions = make(map[string]Action)
+			return &actions, nil
+		}
+		return nil, fmt.Errorf("cannot get IST actions on height %d: %w", height, err)
+	}
+	// decode the IST actions
+	if err := actions.decode(actionsBytes); err != nil {
+		return nil, fmt.Errorf("could not decode actions: %w", err)
+	}
+	return &actions, nil
+}
+
+// Reschedule reschedules an IST action to the given new height.
+func (c *Controller) Reschedule(id []byte, oldHeight, newHeight uint32) error {
+	if id == nil {
+		return fmt.Errorf("cannot reschedule IST action: nil id")
+	}
+	if oldHeight >= newHeight {
+		return fmt.Errorf("cannot reschedule IST action: old height (%d) >= new height (%d)", oldHeight, newHeight)
+	}
+	actions, err := c.Actions(oldHeight)
+	if err != nil {
+		return err
+	}
+	// find the IST action
+	var actID string
+	for actID = range actions.Actions {
+		if string(id) == actID {
+			break
+		}
+	}
+	if actID == "" {
+		return fmt.Errorf("cannot reschedule IST action: action not found")
+	}
+	// store the IST action
+	a := actions.Actions[actID]
+	a.Attempts++
+	if err := c.Schedule(newHeight, id, a); err != nil {
+		return err
+	}
+	// delete the IST action
+	delete(actions.Actions, actID)
+
+	// update the IST actions for the old height
+	if err := c.storeToNoState(dbIndex(oldHeight), actions.encode()); err != nil {
+		return fmt.Errorf("cannot reschedule IST action: %w", err)
+	}
+	return nil
+}
+
+// Commit executes the IST actions scheduled for the given height.
+func (c *Controller) Commit(height uint32, isSynchronizing bool) error {
+	actions, err := c.Actions(height)
+	if err != nil {
+		return fmt.Errorf("cannot commit IST actions: %w", err)
+	}
+	if len(actions.Actions) == 0 {
+		return nil
+	}
+	for id, action := range actions.Actions {
+		switch action.Action {
+		case ActionComputeResults:
+			log.Debugw("compute results", "height", height, "id", fmt.Sprintf("%x", id), "attempt", action.Attempts)
+			// check if we have all the revealed keys, else reschedule
+			if !c.checkRevealedKeys(action.ElectionID) {
+				// we are missing some keys, we cannot compute the results yet
+				newHeight := height + (1 + action.Attempts*2)
+				log.Infow("missing keys, rescheduling IST action", "newHeight", newHeight, "id",
+					fmt.Sprintf("%x", id), "action", ActionsToString[action.Action], "attempt", action.Attempts)
+				if err := c.Reschedule([]byte(id), height, newHeight); err != nil {
+					return fmt.Errorf("cannot reschedule IST action: %w", err)
+				}
+				continue
+			}
+			if !isSynchronizing {
+				// compute results in a goroutine to avoid blocking the state transition
+				go func() {
+					if err := c.computeAndStoreResults(action.ElectionID); err != nil {
```

Comment:
you might have multiple of these calls happen concurrently, would that be safe?

Reply:
Yes, it should be safe.